### DISCUSSION TITLE
Version 0.3.0 development progress

### DISCUSSION THREAD

**Discussion by txoof** [txoof/PaperPi#3] [General]
@aaronr8684
Let's use this thread to keep track of what's going on...

**Comment by txoof** [comment]
@aaronr8684 do you have a recipe for spinning up Pi VMS? I've been following some qemu guides, but struggling to find one that actually explains the steps.

**Reply by aaronr8684** [reply]
I don't because I have so many actual devices laying around that I just image a card when I need to test something. However, when I've done other testing on a VM, I usually will just download a prebuilt image from OSBoxes.com for the given OS I need. I just checked and the latest Raspbian they have is from the Jan 11, 2021. But once you have the VM image up and running, you could let an ansible or shell script to run the commands to get it to the latest release and then save that as a base image to copy for future tests. I'm not sure if automating beyond this is worth the effort.

**Reply by txoof** [reply]
Great! I'll look into the solution from OSBoxes. I also have a bunch of SD cards and a whole mess of PIs laying around and it's pretty easy to just dump a new image on an SD card and fiddle with that.

**Comment by aaronr8684** [comment]
Is there anything you have left to build out or something specific you wanted me to look at or were you just looking for another set of eyes to look over the code in general? I'm assuming you'll want to put the effort into finishing this rework before we focus on building the Omni functionality into it.

**Reply by txoof** [reply]
Yeah, I'd like to get this reworked before we dive into the Omni EPD stuff. Once this is more stable, would be a great time to implement some of the features of OmniEPD into PaperPi.

In the mean time, I can help on the epdlib side.

Right now in this 0.3 version of paperpi, I'm focusing on getting an installer sorted out; it will need some testing to make sure it's sane and I would really love your help with that.

Another thing I'd like to try to figure out is how to do an installer like HomeBrew uses. It's a super amateur-friendly system. They have a clever way of pulling down a bash script with curl and then piping it directly to bash to do all the work without much intervention from the user. You can see the method [here](https://brew.sh/) and the script that does the heavy lifting [here](https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh).

**Would you be interested in trying to sort out how that installer works and adding that to this project?**

**Reply by aaronr8684** [reply]
Yea let me look into it.

**Reply by txoof** [reply]
@aaronr8684,
I think this version is ready for testing. The basics should all be in place, though the process is a little rough around the edges. To test it out, you can clone this repo.

For an executable that runs just from within the cloned repo try: `./utilities/create_devel_environment.sh -c` -- that should create a pipenv with all the proper modules installed. You can cd into the paperpi directory and run `pipenv run paperpi.py`

For the full install experience you can run the following from the root of the repo: `sudo ./install/install.sh` That should install into `/usr/local/bin` and `/usr/local/lib/paperpi`, create the daemon and daemon user.

I've done some limited testing, but another set of eyes and hands trying to break it would be great.

Thanks!

**Reply by aaronr8684** [reply]
As I find things in testing, do you want to track them here or do you want me to open issues that we can use to work through them one by one?

**Reply by txoof** [reply]
I think open issues in this project will be easier to track and manage.

**Reply by txoof** [reply]
I've also got a working homebrew style remote installer working. It pulls the repo and runs the install script in a somewhat automated way.

You'll need to get the RAW link to the [remote_installer.sh](https://github.com/txoof/PaperPi/blob/manage_modules/install/remote_install.sh) because the repo is still private. The command below should run the installer:

`/bin/bash -c "$(curl -fsSL RAW_CONTENT_LINK_HERE)"`

**Reply by aaronr8684** [reply]
Nice. I was looking at another one line installer from a product I came across. It does some architecture checking and consent/license agreement before downloading and calling the install. Your version looks like a great start. I'll test it out.

**Reply by aaronr8684** [reply]
Auto-install seems to work great on a 32-bit Raspbian (once I ironed out the SSH mess for connecting to private repos 😅 ). 

For updates, do you want me to keep submitting PRs or just direct commit to the manage_modules branch? I like the PRs because it gives you the ability to give the code a quick glance before merging. I just don't want to be pinging you with PRs if you don't care.

**Reply by txoof** [reply]
The pr work flow is great so far. 

Once we've ironed out most of the kinks, it would probably be good for me to start doing PRs too so we can benefit from a review process all around.

It's really great to have another person thinking along on this project.

**Reply by aaronr8684** [reply]
Sounds good and I'm glad to help.

**Reply by aaronr8684** [reply]
If you're ready to push to main, I could run one more test on a 32-bit device and 64-bit device just to make sure everything works as intended. We've been testing as we go, so 🤞 it just works.

**Reply by aaronr8684** [reply]
Just picked up two more screens (for various projects) and I selected screens that also would help diversify our testing cases. 

So now I have:
- 7.5b_V2 (red)
- 5.65 (color)
- 4.2c (yellow)
- 3.7 (greyscale-4) - officially supports partial update
- 2.13 (bw) - officially supports partial update

**Reply by txoof** [reply]
I'll get on to tidying things up and get ready to push all the changes into the main branch. Thanks for the house-keeping in the install scripts.

How are things going with the new screens and omni-epd and epdlib?

**Reply by aaronr8684** [reply]
Still waiting on the yellow screen from china. I didn't realize how hard it was to find a yellow waveshare screen. I almost pulled the trigger and ordered a pimoroni screen, but I really wanted to have a WS yellow device to dig more into the Omni lib. I've been testing the partial refresh on the 2.13 and it's pretty slick. I know the 3.7 supports that as well, but that driver file is so unique, I didn't feel like it would be a good test case by itself.

As far as Omni, I've been doing some exploration into the code but had training all week so I couldn't block off enough time to really focus on changing anything. I was able to lay out a roadmap for changes I'd like to implement (in no particular order and not necessarily dependent on the others unless noted)

- [x] Proof of concept for Omni support in EPDLib
- [x] Abstract WS specific code in EPDLib to support other libraries from Omni
- [ ] Add partial update support in Omni (only screens that have official support)
- [x] Use Omni for IT8951 screens in EPDLib (dependent on the next item)
- [x] Wait for IT8951 support in Omni (ongoing)
- [ ] Verify Inky support in EPDLib (outside testers needed)
- [ ] Add greyscale support in EPDLib
- [ ] Add 3 and 7-color support in EPDLib
- [ ] Bring updated EPDLib into PaperPi

I'm trying to bite off these changes into small chunks so I'm not overwhelmed or breaking major items along the way, but I'm enjoying working on it. 

On a side note, have you ever considered splitting off from Jupyter for the codebase? I know you like to develop in there and I can appreciate how easy it is to prototype in a notebook, but it might be easier to maintain across multiple devs if it wasn't in that format. I'll follow your lead (it is your project after all). I mean for me, the worst case is that I develop in the *.py files and then just copy/paste the changes to the Jupyter files before a major merge and it's really not that big of a deal.

**Reply by aaronr8684** [reply]
Also, if you're curious, I compiled all the waveshare screens and their (case sensitive) methods as well as the driver files with resolution info. It's really useful when I need to quickly look up something or compare across multiple screens.

https://docs.google.com/spreadsheets/d/1r4EKPGZvz_xQZA58WPyuTVPhXGKCWd7-Z-btqgaFfKw/edit?usp=sharing

**Reply by txoof** [reply]
> On a side note, have you ever considered splitting off from Jupyter for the codebase? I know you like to develop in there and I can appreciate how easy it is to prototype in a notebook, but it might be easier to maintain across multiple devs if it wasn't in that format. I'll follow your lead (it is your project after all). I mean for me, the worst case is that I develop in the *.py files and then just copy/paste the changes to the Jupyter files before a major merge and it's really not that big of a deal.

I'm really glad you brought this up. When I shared this on Reddit I was really hoping to find some collaborators that would challenge me to think in different ways. I've been working on projects like this alone for a really long time and I'm not very good at challenging myself to think in new ways. Moving out of Jupyter is a big step for me, but I would like to be more flexible in how I work and learn new work flows.

I'm so glad you brought up VS Code the other day. I started using it for editing MarkDown and I really love it. I also started using it at work for a **HELLISH** project that involves learning PL/SQL on the fly and poorly documented XML. Long-story short: I learned what linters are and found a bunch of VS Code extensions that are saving my sanity. 

I really like working in an environment where I can fiddle around with the code and debug with code snips. The flexibility to run code blocks in any order without refactoring is pretty much the killer feature for me. I can come up with an idea, set some variables, run my idea against the variables and interrogate the results right in the notebook. It's really great for me. 

Tell me more about your workflow in VS Code/PyCharm/Whatever. What tools do you use to help you debug and work through issues and test out ideas?

**Reply by txoof** [reply]
> Still waiting on the yellow screen from china. I didn't realize how hard it was to find a yellow waveshare screen. I almost pulled the trigger and ordered a pimoroni screen, but I really wanted to have a WS yellow device to dig more into the Omni lib. I've been testing the partial refresh on the 2.13 and it's pretty slick. I know the 3.7 supports that as well, but that driver file is so unique, I didn't feel like it would be a good test case by itself.

I've been thinking about picking up a pimoroni screen. Which screen would you recommend to facilitate testing in the future? I wouldn't mind picking up a color screen. It would be pretty killer to have a color version of PaperPi on my desk.

**Reply by aaronr8684** [reply]
> I've been thinking about picking up a pimoroni screen. Which screen would you recommend to facilitate testing in the future? I wouldn't mind picking up a color screen. It would be pretty killer to have a color version of PaperPi on my desk.

The pimoroni screens seem to be divided in the pHAT (smaller bw and 3-color), wHAT (med bw and 3-color), and Impression (7-color) categories. If you're looking at color anyway, I would pick up one of the Impression screens. They're pretty cheap for what they are.

They've also reworked their library so the pHAT and wHAT use the same code now, so one of them should cover both categories.

**Reply by aaronr8684** [reply]
> I'm really glad you brought this up. When I shared this on Reddit I was really hoping to find some collaborators that would challenge me to think in different ways. I've been working on projects like this alone for a really long time and I'm not very good at challenging myself to think in new ways. Moving out of Jupyter is a big step for me, but I would like to be more flexible in how I work and learn new work flows.

I'm no expert here, and really just jumped into this to learn how to work on others' code bases. It has been a great learning experience and is nice to see how other's approach a solution. I like Jupyter for rapid prototyping and super quick projects and working on this stuff has reminded me of how nice of a tool it can be.

> I'm so glad you brought up VS Code the other day. I started using it for editing MarkDown and I really love it. I also started using it at work for a **HELLISH** project that involves learning PL/SQL on the fly and poorly documented XML. Long-story short: I learned what linters are and found a bunch of VS Code extensions that are saving my sanity.
 
I really think extension is where VSCode shines. It's not a bad vanilla IDE (and coming from someone who can't stand the full VS software), but the extension really push it to the top of my list. 

> I really like working in an environment where I can fiddle around with the code and debug with code snips. The flexibility to run code blocks in any order without refactoring is pretty much the killer feature for me. I can come up with an idea, set some variables, run my idea against the variables and interrogate the results right in the notebook. It's really great for me.

I need to start using Jupyter more for this. I tend to just create a temp project in PyCharm or VSCode and then play around, but Jupyter would definitely be a great option. PyCharm automatically creates 

> Tell me more about your workflow in VS Code/PyCharm/Whatever. What tools do you use to help you debug and work through issues and test out ideas?

It really depends on what the project is on which one I use. 

For quick projects or full Windows environment, I really like PyCharm. It automatically creates a VENV so you aren't messing with your system libraries or configs and allows each new project to have a fresh start. It also has a lot of convivence features, but unfortunately is just for coding in Python. That means that I usually am using something else for other languages. 

So for times where I'm doing mixed coding (Python and something else), working on Linux/Pi's, or using Git, I'm starting to really like VSCode. I don't get the automatic VENV, but using the Remote SSH extension, I can code on my fast windows desktop and then just upload the changes to a pi for testing instead of trying to do it all via VNC or another monitor. The built in Git support is so nice and works over SSH so all of the files and libraries can live on the Pi and changes can be commited in VSCode up to GitHub. I just discovered the "Save as Root in Remote - SSH" extension that allows saving via sudo that isn't supported natively in the Remote - SSH extension.

**Reply by aaronr8684** [reply]
@txoof What's the plan for pushing people from v0.2.16 (epd_display) to v0.3.0 (PaperPi)?

Also, have you considered renaming EPDLib to something more descriptive of what it will be. I'm not sure the best name, just a thought I had. Maybe something like EPDscreen or EPDwindow? It might be more trouble than it's worth 🤷‍♂️

**Reply by txoof** [reply]
I have two ideas for moving people to this new version once we are ready:
1. mark the other repo as inactive and add a pointer to V2
2. merge this in on top of the previous repo.

**Plan 1**

Upsides:
- ditch the horrendous, huge set of deltas from all the .tgz in the git history
- fresh start with a nice clean tree

Downsides:
- losing people with the a new url

**Plan 2**
Upsides:
- keep the url in a well known place
- maintain history

Downsides:
- keep some pretty terrible history
- sit on top of a huge pile of deltas making the clone process **slow**. This worries me the most with respect to the new installer.

Opinions?

**Reply by txoof** [reply]
> Also, have you considered renaming EPDLib to something more descriptive?

Hmm. This would mean a new pypi entry, but that's pretty small overhead. I doubt many people are relying on it yet so it shouldn't put too many people out.

I'll sleep on it.

**Reply by aaronr8684** [reply]
> Hmm. This would mean a new pypi entry, but that's pretty small overhead. I doubt many people are relying on it yet so it shouldn't put too many people out.

It's such a minor item, if you think it will cause any headaches, I think you could scrap it. Just a random thought.

**Reply by aaronr8684** [reply]
> Downsides:
> 
> * keep some pretty terrible history
> * sit on top of a huge pile of deltas making the clone process **slow**. This worries me the most with respect to the new installer.
> 
> Opinions?

To address the second "slow" issue, maybe you could explore this tool to remove the tgz files - https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository

The bad history would still be an issue, but at least the other issue would be resolved. I think keeping the same link (and renaming?) would be the best approach so the momentum you're building now won't be lost.

**Comment by aaronr8684** [comment]
**To-Do for PR into 'main' branch.**

- [ ] Update 'remote_install.sh' branch variable (PP_GITBRANCH)


I figured we should start keeping track of items that will need to be changed when we PR into main. It shouldn't be a long list, but I'm sure we'll miss them if we don't note them now. We can just edit this post to add items.

**Reply by txoof** [reply]
Thanks for catching that. It was lingering in the back of my mind.

**Comment by aaronr8684** [comment]
I didn't think adding spellcheck would be that involved so sorry for the extra commits on main...anyway, I've moved into a dev branch. I might turn this into a generic branch for badge additions after adding spell check

**Reply by aaronr8684** [reply]
Spellcheck is added and I fixed the existing typos (I googled most of them for verification because I'm just as bad at spelling) and created an exception list. 

To summarize what I did/learned (if you're curious):
- Added a new action via the `PaperPi/.github/workflows/spellcheck.yml` file
- Created a configuration via the `.spellcheck.yml` file
- [Ran the job](https://github.com/txoof/PaperPi/actions/runs/1913927554) and got an output of the errors 
- Used [Google Colab](https://colab.research.google.com/drive/1Qr0qjGkkyVEp1CfDKsS7j4f7Z30hJOF2?usp=sharing) (their Jupyter clone) to get a panda dataframe from the output (not my code)
- Output the summary data as a [CSV](https://drive.google.com/file/d/1rInTBAN1M1ocew6FC6f84CkqH1tZs857/view?usp=sharing)
- Copied the list to the `.wordlist.txt` "don't flag" file and deleted the actual typos
- [Ran the job again](https://github.com/txoof/PaperPi/actions/runs/1914065644)
- Used this much shorter list to fix the typos in the `*.md` files
- [Ran the job one last time](https://github.com/txoof/PaperPi/actions/runs/1914143714)
- [Squashed and merged](https://github.com/txoof/PaperPi/pull/33) the changes to main. If you click on details of the PR, you will see the action running automatically.
- Fixed the badge to point to the main branch on the README.md. This badge will automatically change if the job fails. 

It was a fun learning experience with actions and badges. I figured it would be a good one to get my feet wet and maybe more can be added in the future as you see a need. 

For the action documentation, I mostly used the [GitHub Marketplace Docs](https://github.com/marketplace/actions/github-spellcheck-action) and the repo mentioned in there. 

I used the action's actual badge, but for additional badges/ideas, there are quite a few listed at https://shields.io/

**Reply by aaronr8684** [reply]
Also, we could move the two dot files created in the root to a sub-folder and update the config files to point to the new location if you want the root to stay cleaner on GitHub

**Reply by txoof** [reply]
Wow. This is really neat. I haven't played with any of the automated actions. I'll have to read up in this. 

Does this then validate all the .md (even within the plugins?) for spelling and change the badge when something is wonky?

**Reply by aaronr8684** [reply]
Yep. Here's the list after I added the wordlist that shows file and word misspelled. It also supports doing the same thing in python files and others, but I wasn't ready to tackle everything yet.

[spellcheck_results_after_wordlist.txt](https://github.com/txoof/PaperPi/files/8162025/spellcheck_results_after_wordlist.txt)

**Reply by aaronr8684** [reply]
It also will run the script anytime someone submits a PR to check their readme's as well which is nice

**Reply by aaronr8684** [reply]
@txoof What are your thoughts on these...I'm still undecided

https://github.com/txoof/PaperPi/tree/add_badges#helping-out

**Reply by txoof** [reply]
I like the looks of it. The automation is neat.

Would it make sense to move these closer to the top?

**Reply by aaronr8684** [reply]
I wish they would click through to the actual issue labels. We'll have to look at the list and see which ones make sense to add. One of the actions and badge items that would be nice to get working is the auto push to PyPi. Maybe that doesn't make sense for PaperPi, but it would be nice for EPDLib. Lots of potential all around for automation.

As far as the top, I'm not sure. I think they naturally fit by the help section, but I don't have a strong opinion either way. 

Do you like the look of the first two or second two better?

**Comment by txoof** [comment]
I was hoping to have more time this week to work on stuff, but work is kind of nuts. We have a huge deadline next week and I've been going full out on solving issues.

Hopefully after we pass this milestone, things will settle down a little

**Reply by aaronr8684** [reply]
All good. I've been kind of busy at work as well, but have found a little time to help with the Omni project to get some of the last bugs squashed before moving forward with bringing it over to EPDLib. Hopefully, that will also give him some time to finish the merge with the IT8951 screens. I never imagined I'd know so much about the pillow library 😆

**Reply by txoof** [reply]
The pillow library is really incredible, but suffers from so much accumulated history. I do love how consistent they try to be and their focus on backwards compatibility is admirable!

**Reply by aaronr8684** [reply]
@txoof  I think the EPDLib is at a good point to bring in to EPD_display/PaperPi. Do you want to bring it in here first since it's not used currently and allow us to work out the bugs? 

Right now, EPDLib has the following changes:
- Support for all Omni screens
- Combined logic across all (even HD) screens
- Converted list_compatible_modules to Omni's version

Next phase could include the following:
- tri-color support
- 7-color support
- greyscale support
- partial update support (requires changes in OmniEPD)

If you want to bring at least the color/grey in before making the changes in EPD_display/PaperPi, I can keep working on it. Just let me know. I will need to move the changes over to the ipynb before merging back to main as I've been working in the py files during testing.

**Reply by txoof** [reply]
How does this sound:

Let's work against PaperPi.  I've created a new branch called Omni_epd: https://github.com/txoof/epdlib/tree/omni_epd

Don't worry about back-merging the changes into the jupyter notebooks. We can just work from the .py and I can pull stuff into jupyter as needed to work on it. Fortunately there's a pretty good workflow for bringing straight python into a notebook and breaking it up to fiddle. To simplify everyone's life, let's work primarily in .py files. It will give me a chance to try some other workflows.

Let's prioritize testing against against the omni compatibility and make sure that the config files work and the displays work as expected. I've added another pi to my fleet so I can test on a 5in83, 2in7 and HD all at the same(ish) time without unplugging anything.

Can you PR against the new omni_epd branch at epdlib? We can add that to the deps for PaperPi and then start developing against that.

Once we've got satisfactory compatibility with the new EPDLib, we can start working on the color/gray support.

**Reply by aaronr8684** [reply]
@txoof I'm kinda new to release management stuff, so I might be completely off, but do you want to create a release in the EPDLib (or maybe just use the pypi versioning) that will lock/pin it's current state to v0.5.1.1 for epd_display? Then we can test "latest" or whatever version (v0.6?) we want with PaperPi without maybe breaking some of the epd_display stuff?

**Reply by txoof** [reply]
That's what I had in mind. The current pypy 0.5.1.1 version of epdlib is good enough for epd_display and now we're free to start work on v0.6. We can leave this version set to follow latest.

**Reply by aaronr8684** [reply]
Alright, I think the branches and updates are getting a little spread out on epdlib and paperpi. It looks like some of them are behind others. If you have time, do you think we could get those consolidated into main and/or one or two dev branches per repo? That way, we can make sure we're working on the most up to date code and then run some tests to make sure everything is working as expected still?

I'm thinking for epdlib:
- Main (stable v0.5.1.1)
- Dev (working v0.6.x.x) - I think we can merge what we have in Omni-EPD for now since it's kinda stable and we could do some more testing as we also develop more features in the omni_epd branch
- Omni_EPD - I think this could be our active dev branch for the grey/color support. We can test new code and such. Once we get to a point we feel good, we can merge this to dev to do some more through testing with paperpi before merging to main.

For PaperPi:
- Main (stable v0.3.0) - This is fine for now
- Dev - I think this could be a combination of the epdlib_0.6 and manage_modules branches. This would also be our main branch for testing while referring to the Dev branch in EPDLib until we get feature parity from EPD_Display
- Add_badges - I'm not sure if this will get merged anytime soon, it was more a place for me to explore some workflow and automation. I think it's fine where it is, and I'll keep rebasing as needed to keep it in line with the other branches
- Plugin_installer - Also fine as is for now. There shouldn't be too much overlap between that and the dev branch as far as conflicting changes.

Anyway, I just thought we could simplify these for now. I'm running into a few issues during the test installs and I want to make sure they are actual issues and not just config items that will go away once things are consolidated.

**Reply by txoof** [reply]
Thanks for this. This is exactly the overview I was intending to make today. I'll get started on this.

**Reply by txoof** [reply]
## EPDLib:

- [x] Merged omni_epd branch into development
   - test development branch in freshly built environment against it8951
   - fixed issue with venv build script
- [x] Omni_EPD matches development branch

**Reply by txoof** [reply]
## PaperPi
- [x] merged in manage_modules and epdlib_0.6
   - made utilities scripts a little more robust
   - successfully tested development environment with epdlib 0.6 on IT8951 9" screen; grayscale not working
- [x] deleted manage_modules branch on github
- [x] deleted epdlib_0.6 branch on github

**Reply by txoof** [reply]
should be all good now.

I'll keep plugging away on the plugin installer. Let me know if you need any help with the logic for epdlib grayscale.

**Reply by aaronr8684** [reply]
Sorry, my day job has been a little crazy and I've been working on my daughter's birthday present in the evenings (below), but I should have some more time to work on this again. Also, I've been thinking about discord more and I think it would be good/nice to have a support server available if you're open to it. I'm happy to set it up for you if you want. It would probably be better to create one from scratch instead of reusing the one you shared. What are your thoughts?

![2022-03-24 23 30 21](https://user-images.githubusercontent.com/57102156/160134433-80bb41be-3059-4a31-b6c4-68f05c4ed2b3.jpg)

**Reply by txoof** [reply]
That's a great looking gift. Your daughter is going to love that!

My work has been crazy busy too. We're deploying a new learning platform and it's super intense. I'm learning/writing SQL ath the same time along with all kinds of problem solving

When I'm done for the day at work, my brain is pretty toast. I'm going to spend some time later today patching up a few lingering issues and see if I can't support some of the recent questions.

Let's try out a discord for support. I haven't used discord much, but I'm game.

**Reply by aaronr8684** [reply]
> Let's try out a discord for support. I haven't used discord much, but I'm game.

Go ahead and setup a new server (or we can use the existing, up to you). Add me and allow admin access (you can create an admin role). I'll do the rest
